File: docs/perry-main-src/perry-main/benchmarks/honest_bench/workloads/3_image_convolution/zig/src/main.zig
// 5x5 Gaussian blur on a 3840x2160 RGB image.
// In-memory input + output checksum; see workload README.md for rationale.

const std = @import("std");

const W: usize = 3840;
const H: usize = 2160;
const N: usize = W * H * 3;
const SEED: u32 = 0x9E3779B9;

const KERNEL: [5][5]i32 = .{
    .{ 1, 4, 7, 4, 1 },
    .{ 4, 16, 26, 16, 4 },
    .{ 7, 26, 41, 26, 7 },
    .{ 4, 16, 26, 16, 4 },
    .{ 1, 4, 7, 4, 1 },
};
const KSUM: i32 = 273;

fn generate_input(alloc: std.mem.Allocator) ![]u8 {
    const pixels = try alloc.alloc(u8, N);
    var y: usize = 0;
    while (y < H) : (y += 1) {
        const row_base: u8 = @intCast((y * 255) / H);
        const off0 = y * W * 3;
        var x: usize = 0;
        while (x < W) : (x += 1) {
            const off = off0 + x * 3;
            pixels[off] = @intCast((x * 255) / W);
            pixels[off + 1] = row_base;
            pixels[off + 2] = @intCast(((x + y) * 255) / (W + H));
        }
    }
    var s: u32 = SEED;
    var i: usize = 0;
    while (i + 4 <= N) {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        pixels[i] = pixels[i] +% @as(u8, @truncate(s));
        pixels[i + 1] = pixels[i + 1] +% @as(u8, @truncate(s >> 8));
        pixels[i + 2] = pixels[i + 2] +% @as(u8, @truncate(s >> 16));
        pixels[i + 3] = pixels[i + 3] +% @as(u8, @truncate(s >> 24));
        i += 4;
    }
    while (i < N) {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        pixels[i] = pixels[i] +% @as(u8, @truncate(s));
        i += 1;
    }
    return pixels;
}

inline fn clamp_idx(v: isize, lo: isize, hi: isize) usize {
    if (v < lo) return @intCast(lo);
    if (v > hi) return @intCast(hi);
    return @intCast(v);
}

inline fn clamp_u8(v: i32) u8 {
    if (v < 0) return 0;
    if (v > 255) return 255;
    return @intCast(v);
}

fn blur(alloc: std.mem.Allocator, src: []const u8, w: usize, h: usize) ![]u8 {
    const dst = try alloc.alloc(u8, src.len);
    const wi: isize = @intCast(w);
    const hi: isize = @intCast(h);
    var y: usize = 0;
    while (y < h) : (y += 1) {
        var x: usize = 0;
        while (x < w) : (x += 1) {
            var r_acc: i32 = 0;
            var g_acc: i32 = 0;
            var b_acc: i32 = 0;
            var ky: isize = -2;
            while (ky <= 2) : (ky += 1) {
                const yy = clamp_idx(@as(isize, @intCast(y)) + ky, 0, hi - 1);
                const row = yy * w;
                var kx: isize = -2;
                while (kx <= 2) : (kx += 1) {
                    const xx = clamp_idx(@as(isize, @intCast(x)) + kx, 0, wi - 1);
                    const idx = (row + xx) * 3;
                    const k: i32 = KERNEL[@as(usize, @intCast(ky + 2))][@as(usize, @intCast(kx + 2))];
                    r_acc += @as(i32, src[idx]) * k;
                    g_acc += @as(i32, src[idx + 1]) * k;
                    b_acc += @as(i32, src[idx + 2]) * k;
                }
            }
            const out_idx = (y * w + x) * 3;
            dst[out_idx] = clamp_u8(@divTrunc(r_acc, KSUM));
            dst[out_idx + 1] = clamp_u8(@divTrunc(g_acc, KSUM));
            dst[out_idx + 2] = clamp_u8(@divTrunc(b_acc, KSUM));
        }
    }
    return dst;
}

fn fnv1a32(bytes: []const u8) u32 {
    var h: u32 = 0x811c9dc5;
    for (bytes) |b| {
        h ^= @as(u32, b);
        h *%= 0x01000193;
    }
    return h;
}

pub fn main() !u8 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const src = try generate_input(alloc);
    defer alloc.free(src);
    const dst = try blur(alloc, src, W, H);
    defer alloc.free(dst);
    const hash = fnv1a32(dst);

    var stdout_buf: [128]u8 = undefined;
    const out = try std.fmt.bufPrint(&stdout_buf, "checksum={x:0>8} dims={d}x{d} bytes={d}\n", .{ hash, W, H, dst.len });
    _ = try std.fs.File.stdout().writeAll(out);
    return 0;
}

File: docs/perry-main-src/perry-main/benchmarks/honest_bench/workloads/3_image_convolution/rust/src/main.rs
// 5x5 Gaussian blur on a 3840x2160 RGB image.
//
// Input is generated in-memory (deterministic xorshift32) so the three
// language implementations can compare convolution-loop performance
// independent of their binary-file I/O stories. Output is a 64-bit FNV-1a
// checksum of the blurred pixel bytes printed to stdout.

use std::env;
use std::process::ExitCode;

const W: usize = 3840;
const H: usize = 2160;
const N: usize = W * H * 3;
const SEED: u32 = 0x9E3779B9;

// sigma=1.0 Gaussian, sum = 273
const KERNEL: [[i32; 5]; 5] = [
    [1,  4,  7,  4, 1],
    [4, 16, 26, 16, 4],
    [7, 26, 41, 26, 7],
    [4, 16, 26, 16, 4],
    [1,  4,  7,  4, 1],
];
const KSUM: i32 = 273;

fn generate_input() -> Vec<u8> {
    // Matches scripts/gen_image.py: gradient + xorshift32 noise, summed mod 256.
    let mut pixels = vec![0u8; N];
    for y in 0..H {
        let row_base = ((y * 255) / H) as u8;
        let off0 = y * W * 3;
        for x in 0..W {
            let off = off0 + x * 3;
            pixels[off]     = ((x * 255) / W) as u8;
            pixels[off + 1] = row_base;
            pixels[off + 2] = (((x + y) * 255) / (W + H)) as u8;
        }
    }
    let mut s: u32 = SEED;
    let mut i = 0;
    while i + 4 <= N {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        pixels[i]     = pixels[i].wrapping_add(s as u8);
        pixels[i + 1] = pixels[i + 1].wrapping_add((s >> 8) as u8);
        pixels[i + 2] = pixels[i + 2].wrapping_add((s >> 16) as u8);
        pixels[i + 3] = pixels[i + 3].wrapping_add((s >> 24) as u8);
        i += 4;
    }
    while i < N {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        pixels[i] = pixels[i].wrapping_add(s as u8);
        i += 1;
    }
    pixels
}

#[inline(always)]
fn clamp(v: isize, lo: isize, hi: isize) -> usize {
    if v < lo { lo as usize } else if v > hi { hi as usize } else { v as usize }
}

fn blur(src: &[u8], w: usize, h: usize) -> Vec<u8> {
    let mut dst = vec![0u8; src.len()];
    let wi = w as isize;
    let hi = h as isize;
    for y in 0..h {
        for x in 0..w {
            let mut r_acc: i32 = 0;
            let mut g_acc: i32 = 0;
            let mut b_acc: i32 = 0;
            for ky in -2..=2isize {
                let yy = clamp(y as isize + ky, 0, hi - 1);
                let row = yy * w;
                for kx in -2..=2isize {
                    let xx = clamp(x as isize + kx, 0, wi - 1);
                    let idx = (row + xx) * 3;
                    let k = KERNEL[(ky + 2) as usize][(kx + 2) as usize];
                    r_acc += src[idx]     as i32 * k;
                    g_acc += src[idx + 1] as i32 * k;
                    b_acc += src[idx + 2] as i32 * k;
                }
            }
            let out_idx = (y * w + x) * 3;
            dst[out_idx]     = (r_acc / KSUM).clamp(0, 255) as u8;
            dst[out_idx + 1] = (g_acc / KSUM).clamp(0, 255) as u8;
            dst[out_idx + 2] = (b_acc / KSUM).clamp(0, 255) as u8;
        }
    }
    dst
}

fn fnv1a32(bytes: &[u8]) -> u32 {
    // 32-bit variant. Perry's 64-bit BigInt arithmetic has a live bug
    // (bitwise ops wrap signed → hash collapses to 0), so we keep hashes in
    // the 32-bit integer space where all three languages agree bit-exactly.
    let mut h: u32 = 0x811c9dc5;
    for &b in bytes {
        h ^= b as u32;
        h = h.wrapping_mul(0x01000193);
    }
    h
}

fn main() -> ExitCode {
    // Accept dims override for quick sanity checks: `image_conv <w> <h>`
    let args: Vec<String> = env::args().collect();
    let (w, h) = if args.len() >= 3 {
        (args[1].parse::<usize>().unwrap_or(W), args[2].parse::<usize>().unwrap_or(H))
    } else {
        (W, H)
    };
    // Fast-path: precomputed constants assume default dims. For arbitrary dims,
    // regenerate and adjust.
    let src = if w == W && h == H {
        generate_input()
    } else {
        let mut px = vec![0u8; w * h * 3];
        let mut s: u32 = SEED;
        let mut i = 0;
        while i < px.len() {
            s ^= s << 13; s ^= s >> 17; s ^= s << 5;
            px[i] = s as u8;
            i += 1;
        }
        px
    };
    let out = blur(&src, w, h);
    let hash = fnv1a32(&out);
    println!("checksum={:08x} dims={}x{} bytes={}", hash, w, h, out.len());
    ExitCode::from(0)
}
